<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Mario Failli, Jussi Pananen and Vittorio Fortino" />

<meta name="date" content="2020-02-05" />

<title>Pre-processing needed to compile the available .rda files</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Pre-processing needed to compile the available .rda files</h1>
<h4 class="author">Mario Failli, Jussi Pananen and Vittorio Fortino</h4>
<h4 class="date">2020-02-05</h4>



<style type="text/css">

body{ /* Normal  */
      font-size: 12px;
  }
td {  /* Table  */
  font-size: 10px;
}
h1.title {
  font-size: 30px;
  color: DarkRed;
}
h1 { /* Header 1 */
  font-size: 24px;
  color: DarkBlue;
}
h2 { /* Header 2 */
    font-size: 20px;
  color: DarkBlue;
}
h3 { /* Header 3 */
  font-size: 15px;
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 10px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 10px;
}
</style>
<div id="estimating-disease-relevant-genes-and-tissues-before-compiling-the-tissue-specific-efficacy-scores" class="section level2">
<h2>Estimating disease relevant genes and tissues before compiling the tissue-specific efficacy scores</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(ThETA)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">data</span>(gtexv7_zscore)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">data</span>(ppi_strdb_<span class="dv">700</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">data</span>(dis_vrnts)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="kw">data</span>(disease_tissue_zscores)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="kw">data</span>(centrality_score)</span></code></pre></div>
<p>The following code shows how to retrieve disease-associated genes with a specific confident score (see www.disgenet.org/help for more details) and how to compile z-scores for disease-tissue pairs.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="cf">if</span>(<span class="st">&quot;AnnotationDbi&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {</span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb2-3"><a href="#cb2-3"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;AnnotationDbi&quot;</span>)}</span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="cf">if</span>(<span class="st">&quot;MeSH.db&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {</span>
<span id="cb2-6"><a href="#cb2-6"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb2-7"><a href="#cb2-7"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb2-8"><a href="#cb2-8"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;MeSH.db&quot;</span>)}</span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="kw">library</span>(MeSH.db)</span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="kw">library</span>(AnnotationDbi)</span></code></pre></div>
<p>First, the MeshID corresponding to T2DM are retrieved.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>T2DM_mesh &lt;-<span class="st"> </span>AnnotationDbi<span class="op">::</span><span class="kw">select</span>(MeSH.db,<span class="dt">keys =</span> <span class="st">&#39;Diabetes Mellitus, Type 2&#39;</span>, </span>
<span id="cb3-2"><a href="#cb3-2"></a>                                    <span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">&quot;MESHID&quot;</span>,<span class="st">&quot;MESHTERM&quot;</span>),</span>
<span id="cb3-3"><a href="#cb3-3"></a>                                    <span class="dt">keytype =</span> <span class="st">&quot;MESHTERM&quot;</span>)</span></code></pre></div>
<p>Then, variant genes associated to T2DM are collected from DisGeNET.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">library</span>(RCurl)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">library</span>(XML)</span>
<span id="cb4-3"><a href="#cb4-3"></a>T2DM_genes &lt;-<span class="st"> </span><span class="kw">disease.vrnts</span>(T2DM_mesh<span class="op">$</span>MESHID, <span class="dt">id_type =</span> <span class="st">&#39;mesh&#39;</span>, <span class="dt">min_score =</span> <span class="fl">0.6</span>, <span class="dt">curated =</span> T)</span></code></pre></div>
<p>Once the set of disease-relevant genes is downloaded, then the corresponding set of disease-relevant tissues can be determined (Failli et al. 2019). The parameter <em>top</em> can be used to extend the initial set of disease-relevant genes. It indicates how many genes closely related to the disease-genes must be added.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>T2DM_tiss_zscore &lt;-<span class="st"> </span><span class="kw">dis.rel.tissues</span>(<span class="dt">disease_genes =</span> T2DM_genes<span class="op">$</span>entrez, </span>
<span id="cb5-2"><a href="#cb5-2"></a>                                    <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, <span class="dt">weighted =</span> <span class="ot">TRUE</span>,</span>
<span id="cb5-3"><a href="#cb5-3"></a>                                    <span class="dt">tissue_expr_data =</span> gtexv7_zscore, </span>
<span id="cb5-4"><a href="#cb5-4"></a>                                    <span class="dt">thr =</span> <span class="dv">1</span>, <span class="dt">top=</span><span class="dv">0</span>, <span class="dt">rand =</span> <span class="dv">1000</span>, <span class="dt">verbose =</span> T) <span class="co"># rand = 10000 is highly recommended</span></span></code></pre></div>
<p>The previous code generates the following table of z-scores, indicating significant associations between disease and tissues in GTEx.</p>
<p>After compiling scoring disease-tissue associations, the user can compile the tissue-specific efficacy scores as follow.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="cf">if</span>(<span class="st">&quot;magrittr&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;magrittr&quot;</span>)}</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">library</span>(magrittr)</span>
<span id="cb6-3"><a href="#cb6-3"></a>T2DM_scores &lt;-<span class="st"> </span><span class="kw">tissue.specific.scores</span>(T2DM_genes<span class="op">$</span>entrez, </span>
<span id="cb6-4"><a href="#cb6-4"></a>                                        <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, </span>
<span id="cb6-5"><a href="#cb6-5"></a>                                        <span class="dt">directed_network =</span> F, </span>
<span id="cb6-6"><a href="#cb6-6"></a>                                        <span class="dt">tissue_expr_data =</span> gtexv7_zscore,</span>
<span id="cb6-7"><a href="#cb6-7"></a>                                        <span class="dt">dis_relevant_tissues =</span> T2DM_tiss_zscore<span class="op">$</span>z <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="st">                                                             `</span><span class="dt">names&lt;-</span><span class="st">`</span>(<span class="kw">rownames</span>(T2DM_tiss_zscore)),</span>
<span id="cb6-9"><a href="#cb6-9"></a>                                        <span class="dt">W =</span> centrality_score<span class="op">$</span>borda.disc, <span class="dt">cutoff =</span> <span class="dv">5</span>, <span class="dt">verbose =</span> T)</span></code></pre></div>
<div id="the-following-sections-show-how-to-build-from-scratch-the-pre-compiled-.rda-files-needed-for-the-tissue-specific-score" class="section level4">
<h4>The following sections show how to build from scratch the pre-compiled .rda files needed for the tissue specific score</h4>
<p><strong>Collecting gene expression data per tissue from GTEx</strong></p>
<p>In order to create the .rda file <em>gtexv7_zscore</em> the following R packages need to be installed.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="cf">if</span>(<span class="st">&quot;CePa&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;CePa&quot;</span>)}</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="cf">if</span>(<span class="st">&quot;grex&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;grex&quot;</span>)}</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="kw">library</span>(grex)</span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="kw">library</span>(CePa)</span></code></pre></div>
<p>Then, the function <em>read.gct()</em> is utilized to read the GTEx file.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>gtexv7_median_tpm &lt;-<span class="st"> </span>CePa<span class="op">::</span><span class="kw">read.gct</span>(<span class="st">&#39;../data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct.gz&#39;</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="kw">colnames</span>(gtexv7_median_tpm) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.$&quot;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">colnames</span>(gtexv7_median_tpm))</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="kw">colnames</span>(gtexv7_median_tpm) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.+&quot;</span>,<span class="st">&#39;_&#39;</span>,<span class="kw">colnames</span>(gtexv7_median_tpm)) </span></code></pre></div>
<p>The TPM values are log-transformed and then converted to z-scores by using the <em>tissue.exp</em> function.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>gtexv7_zscore &lt;-<span class="st"> </span><span class="kw">tissue.expr</span>(<span class="kw">log2</span>(gtexv7_median_tpm<span class="op">+</span><span class="dv">1</span>))             <span class="co"># to convert exp vals to z-scores</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="kw">rownames</span>(gtexv7_zscore) &lt;-<span class="st"> </span>grex<span class="op">::</span><span class="kw">cleanid</span>(<span class="kw">rownames</span>(gtexv7_zscore))   <span class="co"># to produce Ensembl IDs.</span></span></code></pre></div>
<p>Moreover, all the Ensembl IDs are mapped to Entrez Gene ID.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>gtexv7_gene_ann &lt;-<span class="st"> </span>grex<span class="op">::</span><span class="kw">grex</span>(<span class="kw">rownames</span>(gtexv7_zscore)) </span>
<span id="cb10-2"><a href="#cb10-2"></a>gtexv7_gene_ann &lt;-<span class="st"> </span>gtexv7_gene_ann[<span class="op">!</span><span class="kw">is.na</span>(gtexv7_gene_ann<span class="op">$</span>entrez_id),] </span>
<span id="cb10-3"><a href="#cb10-3"></a>gtexv7_zscore &lt;-<span class="st"> </span>gtexv7_zscore[gtexv7_gene_ann<span class="op">$</span>ensembl_id,]</span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="kw">rownames</span>(gtexv7_zscore) &lt;-<span class="st"> </span>gtexv7_gene_ann<span class="op">$</span>entrez_id</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="co">#save(gtexv7_zscore,file=&#39;gtexv7_zscore.rda&#39;)</span></span></code></pre></div>
<p><strong>Defining a PPI network from STRINGDB</strong></p>
<p>In order to create the .rda file <em>ppi_strdb_700</em> the following R packages need to be installed.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="cf">if</span>(<span class="st">&quot;STRINGdb&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {</span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</span>
<span id="cb11-3"><a href="#cb11-3"></a>    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a>  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;STRINGdb&quot;</span>)}</span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="cf">if</span>(<span class="st">&quot;igraph&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;igraph&quot;</span>)}</span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="kw">library</span>(STRINGdb)</span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="kw">library</span>(igraph)</span></code></pre></div>
<p>First, the Human Protein-Protein interaction network is selected. Then, genes that are annotated in GTEx and connections (or edges) with a score greater than or equal to 700 are kept.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>string_db &lt;-<span class="st"> </span>STRINGdb<span class="op">$</span><span class="kw">new</span>(<span class="dt">version=</span><span class="st">&quot;10&quot;</span>, <span class="dt">species=</span><span class="dv">9606</span>)</span>
<span id="cb12-2"><a href="#cb12-2"></a>gtexv7_mapped &lt;-<span class="st"> </span>string_db<span class="op">$</span><span class="kw">map</span>(<span class="dt">my_data_frame =</span> gtexv7_gene_ann, <span class="dt">my_data_frame_id_col_names=</span><span class="st">&#39;entrez_id&#39;</span>, <span class="dt">removeUnmappedRows=</span>T)</span>
<span id="cb12-3"><a href="#cb12-3"></a>string_inter &lt;-<span class="st"> </span>string_db<span class="op">$</span><span class="kw">get_interactions</span>(gtexv7_mapped<span class="op">$</span>STRING_id)</span>
<span id="cb12-4"><a href="#cb12-4"></a>string_inter &lt;-<span class="st"> </span>string_inter[string_inter<span class="op">$</span>combined_score<span class="op">&gt;=</span><span class="dv">700</span>,] <span class="co"># to select connections with minimum STRING combined score &gt;= 700</span></span>
<span id="cb12-5"><a href="#cb12-5"></a>idx_from &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="dt">x =</span> string_inter<span class="op">$</span>from, <span class="dt">table =</span> gtexv7_mapped<span class="op">$</span>STRING_id)</span>
<span id="cb12-6"><a href="#cb12-6"></a>idx_to &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="dt">x =</span> string_inter<span class="op">$</span>to, <span class="dt">table =</span> gtexv7_mapped<span class="op">$</span>STRING_id)</span>
<span id="cb12-7"><a href="#cb12-7"></a>ppi_strdb_<span class="dv">700</span> &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">node1=</span>gtexv7_mapped<span class="op">$</span>entrez_id[idx_from], <span class="dt">node2=</span>gtexv7_mapped<span class="op">$</span>entrez_id[idx_to], <span class="dt">weight=</span>string_inter<span class="op">$</span>combined_score<span class="op">/</span><span class="dv">1000</span>,<span class="dt">stringsAsFactors =</span> F)</span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="co">#save(ppi_strdb_700,file=&#39;ppi_strdb_700.rda&#39;)</span></span></code></pre></div>
<p>The PPI compiled from STRINGdb will be a data frame containing a symbolic edge list in the first two columns and connection scores in the third column.</p>
<p><strong>Compiling disease-tissue associations</strong></p>
<p>In order to make the .rda files <em>dis_vrnts</em> and <em>disease_tissue_zscores</em> the following R package <em>ontologyIndex</em> need to be installed. Then, the z-score is compiled for all possible disease-tissue pairs.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="cf">if</span>(<span class="st">&quot;ontologyIndex&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;ontologyIndex&quot;</span>)}</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="kw">library</span>(ontologyIndex)</span>
<span id="cb13-3"><a href="#cb13-3"></a></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co"># select the genes associated with each disease annotated in the OBO file</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>efo.OBO &lt;-<span class="st"> </span><span class="kw">get_OBO</span>(<span class="st">&#39;../data/efo_feb2019.obo&#39;</span>, <span class="dt">extract_tags =</span> <span class="st">&quot;everything&quot;</span>) </span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="co"># remove the tag &quot;EFO:&quot; from each id</span></span>
<span id="cb13-7"><a href="#cb13-7"></a>efo_ids &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&#39;EFO:&#39;</span>,<span class="st">&#39;&#39;</span>, <span class="kw">grep</span>(<span class="st">&#39;EFO:&#39;</span>, efo.OBO<span class="op">$</span>id, <span class="dt">value =</span> T)) </span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="co"># select gene-disease associations from DisGeNET</span></span>
<span id="cb13-9"><a href="#cb13-9"></a>dis_vrnts &lt;-<span class="st"> </span><span class="kw">sapply</span>(efo_ids, <span class="cf">function</span>(x) <span class="kw">disease.vrnts</span>(x, <span class="dt">id_type=</span><span class="st">&#39;efo&#39;</span>,</span>
<span id="cb13-10"><a href="#cb13-10"></a>                                                       <span class="dt">min_score=</span><span class="fl">0.6</span>,</span>
<span id="cb13-11"><a href="#cb13-11"></a>                                                       <span class="dt">curated=</span>F), <span class="dt">simplify =</span> F)  </span>
<span id="cb13-12"><a href="#cb13-12"></a><span class="co"># define the toal set of genes from the PPI</span></span>
<span id="cb13-13"><a href="#cb13-13"></a>all_ppi_genes &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">as.character</span>(<span class="kw">unlist</span>(ppi_strdb_<span class="dv">700</span>[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]))) </span>
<span id="cb13-14"><a href="#cb13-14"></a><span class="co"># select the diseases with at least 5 (associated) genes</span></span>
<span id="cb13-15"><a href="#cb13-15"></a>dis_vrnts &lt;-<span class="st">  </span>dis_vrnts[<span class="kw">sapply</span>(dis_vrnts, <span class="cf">function</span>(x)  </span>
<span id="cb13-16"><a href="#cb13-16"></a>                        <span class="kw">length</span>(<span class="kw">intersect</span>(x<span class="op">$</span>entrez,all_ppi_genes)))<span class="op">&gt;=</span><span class="dv">5</span>] </span>
<span id="cb13-17"><a href="#cb13-17"></a><span class="co"># comiple the disease relevant tissues for each selected disease</span></span>
<span id="cb13-18"><a href="#cb13-18"></a>disease_tissue_zscores &lt;-<span class="st"> </span><span class="kw">list.dis.rel.tissues</span>(<span class="dt">disease_gene_list =</span> <span class="kw">sapply</span>(dis_vrnts,<span class="st">&#39;[[&#39;</span>,<span class="dv">1</span>), </span>
<span id="cb13-19"><a href="#cb13-19"></a>                                               <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, </span>
<span id="cb13-20"><a href="#cb13-20"></a>                                               <span class="dt">weighted =</span> <span class="ot">TRUE</span>, </span>
<span id="cb13-21"><a href="#cb13-21"></a>                                               <span class="dt">tissue_expr_data =</span> gtexv7_zscore, <span class="dt">thr =</span> <span class="dv">1</span>, </span>
<span id="cb13-22"><a href="#cb13-22"></a>                                               <span class="dt">top =</span> <span class="dv">0</span>, <span class="dt">rand =</span> <span class="dv">10000</span>, <span class="dt">parallel =</span> <span class="dv">10</span>)</span>
<span id="cb13-23"><a href="#cb13-23"></a><span class="co">#save(dis_vrnts, file=&#39;dis_vrnts.rda&#39;)</span></span>
<span id="cb13-24"><a href="#cb13-24"></a><span class="co">#save(disease_tissue_zscores, file=&#39;disease_tissue_zscores.rda&#39;)</span></span></code></pre></div>
<p><strong>Compiling node centrality scores</strong></p>
<p>The .rda file <em>centrality_score</em> can be compiled by using the R function <em>node.centrality</em>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>centrality_score &lt;-<span class="st"> </span><span class="kw">node.centrality</span>(<span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, </span>
<span id="cb14-2"><a href="#cb14-2"></a>                                    <span class="dt">tissue_expr_data =</span> gtexv7_zscore, </span>
<span id="cb14-3"><a href="#cb14-3"></a>                                    <span class="dt">agg_function =</span> <span class="dv">4</span>, <span class="dt">directed_network=</span>F, </span>
<span id="cb14-4"><a href="#cb14-4"></a>                                    <span class="dt">parallel =</span> <span class="dv">10</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb14-5"><a href="#cb14-5"></a><span class="co">#save(centrality_score,file = &#39;centrality_score.rda&#39;)</span></span></code></pre></div>
<p><strong>Compiling tissue-specifc efficacy scores for all target-disease associations</strong></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>idx &lt;-<span class="st"> </span><span class="kw">apply</span>(disease_tissue_zscores<span class="op">$</span>z,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">any</span>(x <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.6</span>))</span>
<span id="cb15-2"><a href="#cb15-2"></a>disg =<span class="st"> </span><span class="kw">sapply</span>(dis_vrnts[idx],<span class="st">&#39;[[&#39;</span>,<span class="dv">1</span>)   <span class="co"># list of disease-gene sets</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>disr =<span class="st"> </span>disease_tissue_zscores<span class="op">$</span>z[idx,]  <span class="co"># matrix of disease-relevant tissue scores</span></span>
<span id="cb15-4"><a href="#cb15-4"></a>tissue_score &lt;-<span class="st"> </span><span class="kw">list.tissue.specific.scores</span>(<span class="dt">disease_gene_list =</span> disg, </span>
<span id="cb15-5"><a href="#cb15-5"></a>                                            <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, <span class="dt">directed_network =</span> F,</span>
<span id="cb15-6"><a href="#cb15-6"></a>                                            <span class="dt">tissue_expr_data =</span> gtexv7_zscore,  </span>
<span id="cb15-7"><a href="#cb15-7"></a>                                            <span class="dt">dis_relevant_tissues =</span> disr, </span>
<span id="cb15-8"><a href="#cb15-8"></a>                                            <span class="dt">W =</span> centrality_score<span class="op">$</span>borda.disc, <span class="dt">cutoff =</span> <span class="fl">1.6</span>,</span>
<span id="cb15-9"><a href="#cb15-9"></a>                                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>, <span class="dt">parallel =</span> <span class="dv">20</span>)</span></code></pre></div>
</div>
<div id="how-to-format-averaged-tissue-specifc-efficacy-scores" class="section level3">
<h3>How to format averaged tissue-specifc efficacy scores</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>avg_tissue_score &lt;-<span class="st"> </span><span class="kw">sapply</span>(tissue_score,<span class="st">&#39;[[&#39;</span>,<span class="st">&#39;avg_tissue_score&#39;</span>) <span class="op">%&gt;%</span><span class="st"> `</span><span class="dt">rownames&lt;-</span><span class="st">`</span>(<span class="kw">rownames</span>(tissue_score[[<span class="dv">1</span>]]))</span>
<span id="cb16-2"><a href="#cb16-2"></a>avg_tissue_score &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw">melt</span>(avg_tissue_score)</span>
<span id="cb16-3"><a href="#cb16-3"></a>avg_tissue_score[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">lapply</span>(avg_tissue_score[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],  as.character)</span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="kw">colnames</span>(avg_tissue_score) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;target.entrez&#39;</span>,<span class="st">&#39;disease.id&#39;</span>,<span class="st">&#39;avg_tissue.score&#39;</span>)</span>
<span id="cb16-5"><a href="#cb16-5"></a>annotation &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">disease.id=</span>efo.OBO<span class="op">$</span>id,<span class="dt">disease.name=</span>efo.OBO<span class="op">$</span>name)</span>
<span id="cb16-6"><a href="#cb16-6"></a><span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb16-7"><a href="#cb16-7"></a>avg_tissue_score &lt;-<span class="st"> </span><span class="kw">merge</span>(annotation,avg_tissue_score,<span class="dt">by=</span><span class="st">&#39;disease.id&#39;</span>,<span class="dt">all.y=</span>T)</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
