<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Mario Failli &amp; Vittorio Fortino" />

<meta name="date" content="2019-11-18" />

<title>Pre-processing needed to compile the available .rda files</title>

<script src="data:application/x-javascript;base64,JChkb2N1bWVudCkucmVhZHkoZnVuY3Rpb24oKXsKICAgIGlmICh0eXBlb2YgJCgnW2RhdGEtdG9nZ2xlPSJ0b29sdGlwIl0nKS50b29sdGlwID09PSAnZnVuY3Rpb24nKSB7CiAgICAgICAgJCgnW2RhdGEtdG9nZ2xlPSJ0b29sdGlwIl0nKS50b29sdGlwKCk7CiAgICB9CiAgICBpZiAoJCgnW2RhdGEtdG9nZ2xlPSJwb3BvdmVyIl0nKS5wb3BvdmVyID09PSAnZnVuY3Rpb24nKSB7CiAgICAgICAgJCgnW2RhdGEtdG9nZ2xlPSJwb3BvdmVyIl0nKS5wb3BvdmVyKCk7CiAgICB9Cn0pOwo="></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Pre-processing needed to compile the available .rda files</h1>
<h4 class="author">Mario Failli &amp; Vittorio Fortino</h4>
<h4 class="date">2019-11-18</h4>



<style type="text/css">

body{ /* Normal  */
      font-size: 12px;
  }
td {  /* Table  */
  font-size: 10px;
}
h1.title {
  font-size: 30px;
  color: DarkRed;
}
h1 { /* Header 1 */
  font-size: 24px;
  color: DarkBlue;
}
h2 { /* Header 2 */
    font-size: 20px;
  color: DarkBlue;
}
h3 { /* Header 3 */
  font-size: 15px;
  font-family: "Times New Roman", Times, serif;
  color: DarkBlue;
}
code.r{ /* Code block */
    font-size: 10px;
}
pre { /* Code block - determines code spacing between lines */
    font-size: 10px;
}
</style>
<div id="estimating-disease-relevant-genes-and-tissues-before-compiling-the-tissue-specific-efficacy-scores" class="section level2">
<h2>Estimating disease relevant genes and tissues before compiling the tissue-specific efficacy scores</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ThETA)
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Registered S3 method overwritten by 'enrichplot':</span>
<span class="co">#&gt;   method               from</span>
<span class="co">#&gt;   fortify.enrichResult DOSE</span>
<span class="kw">data</span>(gtexv7_zscore)
<span class="kw">data</span>(ppi_strdb_<span class="dv">700</span>)
<span class="kw">data</span>(dis_vrnts)
<span class="kw">data</span>(disease_tissue_zscores)
<span class="kw">data</span>(centrality_score)</code></pre></div>
<p>The following code shows how to retrieve disease-associated genes with a specific confident score (see www.disgenet.org/help for more details) and how to compile z-scores for disease-tissue pairs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span>(<span class="st">&quot;AnnotationDbi&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;AnnotationDbi&quot;</span>)}
<span class="cf">if</span>(<span class="st">&quot;MeSH.db&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;MeSH.db&quot;</span>)}
<span class="kw">library</span>(MeSH.db)
<span class="kw">library</span>(AnnotationDbi)</code></pre></div>
<p>First, the MeshID corresponding to T2DM are retrieved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">T2DM_mesh &lt;-<span class="st"> </span>AnnotationDbi<span class="op">::</span><span class="kw">select</span>(MeSH.db,<span class="dt">keys =</span> <span class="st">'Diabetes Mellitus, Type 2'</span>, 
                                    <span class="dt">columns =</span> <span class="kw">c</span>(<span class="st">&quot;MESHID&quot;</span>,<span class="st">&quot;MESHTERM&quot;</span>),
                                    <span class="dt">keytype =</span> <span class="st">&quot;MESHTERM&quot;</span>)</code></pre></div>
<p>Then, variant genes associated to T2DM are collected from DisGeNET.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RCurl)
<span class="co">#&gt; Loading required package: bitops</span>
<span class="kw">library</span>(XML)
T2DM_genes &lt;-<span class="st"> </span><span class="kw">disease.vrnts</span>(T2DM_mesh<span class="op">$</span>MESHID, <span class="dt">id_type =</span> <span class="st">'mesh'</span>, <span class="dt">min_score =</span> <span class="fl">0.6</span>, <span class="dt">curated =</span> T)</code></pre></div>
<p>Once the set of disease-relevant genes is downloaded, then the corresponding set of disease-relevant tissues can be determined (Failli et al. 2019). The parameter <em>top</em> can be used to extend the initial set of disease-relevant genes. It indicates how many genes closely related to the disease-genes must be added.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">T2DM_tiss_zscore &lt;-<span class="st"> </span><span class="kw">dis.rel.tissues</span>(<span class="dt">disease_genes =</span> T2DM_genes<span class="op">$</span>entrez, 
                                      <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, <span class="dt">weighted =</span> <span class="ot">TRUE</span>,
                                      <span class="dt">tissue_expr_data =</span> gtexv7_zscore, 
                                      <span class="dt">thr =</span> <span class="dv">1</span>, <span class="dt">top=</span><span class="dv">0</span>, <span class="dt">rand =</span> <span class="dv">1000</span>, <span class="dt">verbose =</span> T) <span class="co"># rand = 10000 is higlhy recommended</span></code></pre></div>
<p>The previous code generates the following table of z-scores, indicating significant associations between disease and tissues in GTEx.</p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
maxc
</th>
<th style="text-align:right;">
mer
</th>
<th style="text-align:right;">
sdr
</th>
<th style="text-align:right;">
z
</th>
<th style="text-align:right;">
genes
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Adipose_Subcutaneous
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1.665
</td>
<td style="text-align:right;">
0.7895558
</td>
<td style="text-align:right;">
0.4242892
</td>
<td style="text-align:right;">
15
</td>
</tr>
<tr>
<td style="text-align:left;">
Adipose_Visceral_Omentum
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1.234
</td>
<td style="text-align:right;">
0.5054696
</td>
<td style="text-align:right;">
1.5154225
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
Adrenal_Gland
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1.525
</td>
<td style="text-align:right;">
0.6259127
</td>
<td style="text-align:right;">
0.7588917
</td>
<td style="text-align:right;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
Artery_Aorta
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1.935
</td>
<td style="text-align:right;">
1.3974003
</td>
<td style="text-align:right;">
0.7621295
</td>
<td style="text-align:right;">
21
</td>
</tr>
<tr>
<td style="text-align:left;">
Artery_Coronary
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1.350
</td>
<td style="text-align:right;">
0.5655261
</td>
<td style="text-align:right;">
-0.6188927
</td>
<td style="text-align:right;">
10
</td>
</tr>
</tbody>
</table>
<p>After compiling scoring disease-tissue associations, the user can compile the tissue-specific efficacy scores as follow.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span>(<span class="st">&quot;magrittr&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;magrittr&quot;</span>)}
<span class="kw">library</span>(magrittr)
T2DM_scores &lt;-<span class="st"> </span><span class="kw">tissue.specific.scores</span>(T2DM_genes<span class="op">$</span>entrez, 
                                        <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, 
                                        <span class="dt">directed_network =</span> F, 
                                        <span class="dt">tissue_expr_data =</span> gtexv7_zscore,
                                        <span class="dt">dis_relevant_tissues =</span> T2DM_tiss_zscore<span class="op">$</span>z <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">                                                             `</span><span class="dt">names&lt;-</span><span class="st">`</span>(<span class="kw">rownames</span>(T2DM_tiss_zscore)),
                                        <span class="dt">W =</span> centrality_score<span class="op">$</span>borda.disc, <span class="dt">cutoff =</span> <span class="dv">5</span>, <span class="dt">verbose =</span> T)</code></pre></div>
<div id="the-following-sections-show-how-to-build-from-scratch-the-pre-compiled-.rda-files-needed-for-the-tissue-specific-score" class="section level4">
<h4>The following sections show how to build from scratch the pre-compiled .rda files needed for the tissue specific score</h4>
<p><strong>Collecting gene expression data per tissue from GTEx</strong></p>
<p>In order to create the .rda file <em>gtexv7_zscore</em> the following R packages need to be installed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span>(<span class="st">&quot;CePa&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;CePa&quot;</span>)}
<span class="cf">if</span>(<span class="st">&quot;grex&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;grex&quot;</span>)}
<span class="kw">library</span>(grex)
<span class="kw">library</span>(CePa)</code></pre></div>
<p>Then, the function <em>read.gct()</em> is utilized to read the GTEx file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtexv7_median_tpm &lt;-<span class="st"> </span>CePa<span class="op">::</span><span class="kw">read.gct</span>(<span class="st">'../data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct.gz'</span>)
<span class="kw">colnames</span>(gtexv7_median_tpm) &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.$&quot;</span>,<span class="st">''</span>,<span class="kw">colnames</span>(gtexv7_median_tpm))
<span class="kw">colnames</span>(gtexv7_median_tpm) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.+&quot;</span>,<span class="st">'_'</span>,<span class="kw">colnames</span>(gtexv7_median_tpm)) </code></pre></div>
<p>The TPM values are log-transformed and then converted to z-scores by using the <em>tissue.exp</em> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtexv7_zscore &lt;-<span class="st"> </span><span class="kw">tissue.expr</span>(<span class="kw">log2</span>(gtexv7_median_tpm<span class="op">+</span><span class="dv">1</span>))             <span class="co"># to convert exp vals to z-scores</span>
<span class="kw">rownames</span>(gtexv7_zscore) &lt;-<span class="st"> </span>grex<span class="op">::</span><span class="kw">cleanid</span>(<span class="kw">rownames</span>(gtexv7_zscore))   <span class="co"># to produce Ensembl IDs.</span></code></pre></div>
<p>Moreover, all the Ensembl IDs are mapped to Entrez Gene ID.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtexv7_gene_ann &lt;-<span class="st"> </span>grex<span class="op">::</span><span class="kw">grex</span>(<span class="kw">rownames</span>(gtexv7_zscore)) 
gtexv7_gene_ann &lt;-<span class="st"> </span>gtexv7_gene_ann[<span class="op">!</span><span class="kw">is.na</span>(gtexv7_gene_ann<span class="op">$</span>entrez_id),] 
gtexv7_zscore &lt;-<span class="st"> </span>gtexv7_zscore[gtexv7_gene_ann<span class="op">$</span>ensembl_id,]
<span class="kw">rownames</span>(gtexv7_zscore) &lt;-<span class="st"> </span>gtexv7_gene_ann<span class="op">$</span>entrez_id
<span class="co">#save(gtexv7_zscore,file='gtexv7_zscore.rda')</span></code></pre></div>
<p><strong>Defining a PPI network from STRINGDB</strong></p>
<p>In order to create the .rda file <em>ppi_strdb_700</em> the following R packages need to be installed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span>(<span class="st">&quot;STRINGdb&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
  BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;STRINGdb&quot;</span>)}
<span class="cf">if</span>(<span class="st">&quot;igraph&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;igraph&quot;</span>)}
<span class="kw">library</span>(STRINGdb)
<span class="kw">library</span>(igraph)</code></pre></div>
<p>First, the Human Protein-Protein interaction network is selected. Then, genes that are annotated in GTEx and connections (or edges) with a score greater than or equal to 700 are kept.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">string_db &lt;-<span class="st"> </span>STRINGdb<span class="op">$</span><span class="kw">new</span>(<span class="dt">version=</span><span class="st">&quot;10&quot;</span>, <span class="dt">species=</span><span class="dv">9606</span>)
gtexv7_mapped &lt;-<span class="st"> </span>string_db<span class="op">$</span><span class="kw">map</span>(<span class="dt">my_data_frame =</span> gtexv7_gene_ann, <span class="dt">my_data_frame_id_col_names=</span><span class="st">'entrez_id'</span>, <span class="dt">removeUnmappedRows=</span>T)
string_inter &lt;-<span class="st"> </span>string_db<span class="op">$</span><span class="kw">get_interactions</span>(gtexv7_mapped<span class="op">$</span>STRING_id)
string_inter &lt;-<span class="st"> </span>string_inter[string_inter<span class="op">$</span>combined_score<span class="op">&gt;=</span><span class="dv">700</span>,] <span class="co"># to select connections with minimum STRING combined score &gt;= 700</span>
idx_from &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="dt">x =</span> string_inter<span class="op">$</span>from, <span class="dt">table =</span> gtexv7_mapped<span class="op">$</span>STRING_id)
idx_to &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="dt">x =</span> string_inter<span class="op">$</span>to, <span class="dt">table =</span> gtexv7_mapped<span class="op">$</span>STRING_id)
ppi_strdb_<span class="dv">700</span> &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">node1=</span>gtexv7_mapped<span class="op">$</span>entrez_id[idx_from], <span class="dt">node2=</span>gtexv7_mapped<span class="op">$</span>entrez_id[idx_to], <span class="dt">weight=</span>string_inter<span class="op">$</span>combined_score<span class="op">/</span><span class="dv">1000</span>,<span class="dt">stringsAsFactors =</span> F)
<span class="co">#save(ppi_strdb_700,file='ppi_strdb_700.rda')</span></code></pre></div>
<p>The PPI compiled from STRINGdb will be a data frame containing a symbolic edge list in the first two columns and connection scores in the third column.</p>
<p><strong>Compiling disease-tissue associations</strong></p>
<p>In order to make the .rda files <em>dis_vrnts</em> and <em>disease_tissue_zscores</em> the following R package <em>ontologyIndex</em> need to be installed. Then, the z-score is compiled for all possible disease-tissue pairs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span>(<span class="st">&quot;ontologyIndex&quot;</span> <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>()) <span class="op">==</span><span class="st"> </span><span class="ot">FALSE</span>) {<span class="kw">install.packages</span>(<span class="st">&quot;ontologyIndex&quot;</span>)}
<span class="kw">library</span>(ontologyIndex)

<span class="co"># select the genes associated with each disease annotated in the OBO file</span>
efo.OBO &lt;-<span class="st"> </span><span class="kw">get_OBO</span>(<span class="st">'../data/efo_feb2019.obo'</span>, <span class="dt">extract_tags =</span> <span class="st">&quot;everything&quot;</span>) 
<span class="co"># remove the tag &quot;EFO:&quot; from each id</span>
efo_ids &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'EFO:'</span>,<span class="st">''</span>, <span class="kw">grep</span>(<span class="st">'EFO:'</span>, efo.OBO<span class="op">$</span>id, <span class="dt">value =</span> T)) 
<span class="co"># select gene-disease associations from DisGeNET</span>
dis_vrnts &lt;-<span class="st"> </span><span class="kw">sapply</span>(efo_ids, <span class="cf">function</span>(x) <span class="kw">disease.vrnts</span>(x, <span class="dt">id_type=</span><span class="st">'efo'</span>,
                                                       <span class="dt">min_score=</span><span class="fl">0.6</span>,
                                                       <span class="dt">curated=</span>F), <span class="dt">simplify =</span> F)  
<span class="co"># define the toal set of genes from the PPI</span>
all_ppi_genes &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">as.character</span>(<span class="kw">unlist</span>(ppi_strdb_<span class="dv">700</span>[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]))) 
<span class="co"># select the diseases with at least 5 (associated) genes</span>
dis_vrnts &lt;-<span class="st">  </span>dis_vrnts[<span class="kw">sapply</span>(dis_vrnts, <span class="cf">function</span>(x)  
                        <span class="kw">length</span>(<span class="kw">intersect</span>(x<span class="op">$</span>entrez,all_ppi_genes)))<span class="op">&gt;=</span><span class="dv">5</span>] 
<span class="co"># comiple the disease relevant tissues for each selected disease</span>
disease_tissue_zscores &lt;-<span class="st"> </span><span class="kw">list.dis.rel.tissues</span>(<span class="dt">disease_gene_list =</span> <span class="kw">sapply</span>(dis_vrnts,<span class="st">'[['</span>,<span class="dv">1</span>), 
                                               <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, 
                                               <span class="dt">weighted =</span> <span class="ot">TRUE</span>, 
                                               <span class="dt">tissue_expr_data =</span> gtexv7_zscore, <span class="dt">thr =</span> <span class="dv">1</span>, 
                                               <span class="dt">top =</span> <span class="dv">0</span>, <span class="dt">rand =</span> <span class="dv">10000</span>, <span class="dt">parallel =</span> <span class="dv">10</span>)
<span class="co">#save(dis_vrnts, file='dis_vrnts.rda')</span>
<span class="co">#save(disease_tissue_zscores, file='disease_tissue_zscores.rda')</span></code></pre></div>
<p><strong>Compiling node centrality scores</strong></p>
<p>The .rda file <em>centrality_score</em> can be compiled by using the R function <em>node.centrality</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">centrality_score &lt;-<span class="st"> </span><span class="kw">node.centrality</span>(<span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, 
                                    <span class="dt">tissue_expr_data =</span> gtexv7_zscore, 
                                    <span class="dt">agg_function =</span> <span class="dv">4</span>, <span class="dt">directed_network=</span>F, 
                                    <span class="dt">parallel =</span> <span class="dv">10</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
<span class="co">#save(centrality_score,file = 'centrality_score.rda')</span></code></pre></div>
<p><strong>Compiling tissue-specifc efficacy scores for all target-disease associations</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">idx &lt;-<span class="st"> </span><span class="kw">apply</span>(disease_tissue_zscores<span class="op">$</span>z,<span class="dv">1</span>,<span class="cf">function</span>(x) <span class="kw">any</span>(x <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.6</span>))

tissue_score &lt;-<span class="st"> </span><span class="kw">list.tissue.specific.scores</span>(<span class="dt">disease_gene_list =</span> <span class="kw">sapply</span>(dis_vrnts[idx],<span class="st">'[['</span>,<span class="dv">1</span>) , 
                                            <span class="dt">ppi_network =</span> ppi_strdb_<span class="dv">700</span>, <span class="dt">directed_network =</span> F,
                                            <span class="dt">tissue_expr_data =</span> gtexv7_zscore,  
                                            <span class="dt">dis_relevant_tissues =</span> disease_tissue_zscores<span class="op">$</span>z[idx,], 
                                            <span class="dt">W =</span> centrality_score<span class="op">$</span>borda.disc, <span class="dt">cutoff =</span> <span class="fl">1.6</span>,
                                            <span class="dt">verbose =</span> <span class="ot">FALSE</span>, <span class="dt">parallel =</span> <span class="dv">30</span>)</code></pre></div>
</div>
<div id="how-to-format-averaged-tissue-specifc-efficacy-scores" class="section level3">
<h3>How to format averaged tissue-specifc efficacy scores</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">avg_tissue_score &lt;-<span class="st"> </span><span class="kw">sapply</span>(tissue_score,<span class="st">'[['</span>,<span class="st">'avg_tissue_score'</span>) <span class="op">%&gt;%</span><span class="st"> `</span><span class="dt">rownames&lt;-</span><span class="st">`</span>(<span class="kw">rownames</span>(tissue_score[[<span class="dv">1</span>]]))
avg_tissue_score &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw">melt</span>(avg_tissue_score)
avg_tissue_score[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">lapply</span>(avg_tissue_score[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],  as.character)
<span class="kw">colnames</span>(avg_tissue_score) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'target.entrez'</span>,<span class="st">'disease.id'</span>,<span class="st">'avg_tissue.score'</span>)
annotation &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">disease.id=</span>efo.OBO<span class="op">$</span>id,<span class="dt">disease.name=</span>efo.OBO<span class="op">$</span>name)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
avg_tissue_score &lt;-<span class="st"> </span><span class="kw">merge</span>(annotation,avg_tissue_score,<span class="dt">by=</span><span class="st">'disease.id'</span>,<span class="dt">all.y=</span>T)</code></pre></div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
